File: trees/bst/kth_smallest_element_in_a_bst.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# Approach 1: With global
# @param {TreeNode} root
# @param {Integer} k
# @return {Integer}
def kth_smallest(root, k)
    return root if !root
    @k, @val = k, root.val
    dfs(root)

    @val
end

def dfs(node)
    return if !node
    dfs(node.left)
    @k -= 1
    if @k.zero?
        @val = node.val
        return
    end
    dfs(node.right)
end


# Approach 2: With array
def kth_smallest(root, k)
    inorder(root, [])[k - 1]
end

def inorder(root, a)
    return a if root.nil?
    inorder(root.left, a)
    a << root.val
    inorder(root.right, a)
    a
end

File: trees/bst/kth_smallest_element_in_a_bst.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        self.kth_val = None
        self.k = k
        self.dfs(root)
        return self.kth_val

    def dfs(self, node):
        if not node: return

        self.dfs(node.left)

        self.k -= 1
        if self.k == 0:
            self.kth_val = node.val
            return

        self.dfs(node.right)


# root = [3,1,4,null,2], k = 1

# 230. Kth Smallest Element in a BST
# https://leetcode.com/problems/kth-smallest-element-in-a-bst/description/